packages/inter-protocol/src/proposals/replaceElectorate.js

```
@@ -214,6 +222,15 @@ const startNewEconomicCommittee = async (
   trace(`committeeName ${committeeName}`);
   trace(`committeeSize ${committeeSize}`);
 
+  // Record the retired electorate vat so we can manage it later.
+  const contractInstanceMap = makeScalarBigMapStore(
+    'retiredContractInstances',
+    { durable: true },
+  );
+  const econeconomicCommitteeOriginal = await economicCommitteeOriginalP;
+  contractInstanceMap.init('electorate-v24', econeconomicCommitteeOriginal);
+  retiredContractInstances.resolve(contractInstanceMap);
```

What's the frequency (or "per-ness", as Chip would say) of this resolution? Relatedly, how frequently does this `replaceElectorate.js` proposal get run? I don't see a `provide` pattern here, so I'm worried that this might work correctly during upgrade18, but the *next* time this proposal is run, we'll create a lame-duck `constractInstanceMap`, put something important in it, and then the `retiredContractInstances.resolve()` will be ignored because that key is already in the promise-space. That would lose a reference.

I don't know how to square a provide-pattern with the promise space. Maybe there should be a separate (effectively idempotent) function whose only job is to create `contractInstanceMap` and resolve it into the promise space. Then the electorate replacement function would only `consume` (and not `produce`), and we wouldn't run the risk of this ignored-duplicate-resolve problem.

---

Whatever *next* time is, it'll have to `reset()` in order to re-assign the Map. This is brittle, and will need to be modified the next time it's used, but that isn't the failure mode.

> Then the electorate replacement function would only consume (and not produce), and we wouldn't run the risk of this ignored-duplicate-resolve problem.

I'd be happier if I knew how to do that, but I couldn't think of a way. If that's more important than getting this in tonight, I'll talk to @dckc about how to do that tomorrow.

---

@dckc reminded me of the standard pattern for this. Each intended writer creates a MapStore and attempts to save it as `retiredContractInstances`. One of them will win, the others will be ignored. Each then consumes `retiredContractInstances`, and adds a value to it. They'll all consume the same thing, and they won't know or care whether they were the one that won the first race.

---

I'll defer to his advice, and I also don't know how to code it the way I suggested (it's in the "looks easy because I haven't tried" category :), but that pattern doesn't sound great to me. "The others will be ignored" might be reliable, but it relies upon non-obvious behavior of the promise-space `resolve`, which makes it less legible and an ongoing source of questions ("why read this back out of `consume` when you just put it in there on the previous line?" and "why doesn't this duplicate `resolve` throw?").

---

>  "The others will be ignored" might be reliable, but it relies upon non-obvious behavior of the promise-space resolve, which makes it less legible and an ongoing source of questions ("why read this back out of consume when you just put it in there on the previous line?" and "why doesn't this duplicate resolve throw?").

I agree that it's less than obvious, but it's reliable.

If we ever change our mind about "silently ignore attempts to assign values that are already set", we'll have to change a large number of other things.